homeassistant/components/sensor/recorder.py

```diff
@@ -0,0 +1,96 @@
+"""Statistics helper for sensor."""
+from __future__ import annotations
+
+import datetime
+import logging
+import statistics
+
+from homeassistant.components import history
+from homeassistant.const import ATTR_DEVICE_CLASS
+from homeassistant.core import HomeAssistant
+from homeassistant.util.async_ import run_callback_threadsafe
+
+from . import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+DEVICE_CLASSES = {"temperature"}
+DEVICE_CLASS_STATISTICS = {"temperature": {"mean", "min", "max"}}
+
+
+def _get_entities(hass: HomeAssistant) -> list[tuple[str, str]]:
+    """Get (entity_id, device_class) of all sensors for which to compile statistics."""
+    all_sensors = hass.states.async_entity_ids(DOMAIN)
+    entity_ids = []
+
+    for entity_id in all_sensors:
+        state = hass.states.get(entity_id)
+        if not state:
+            continue
+        device_class = state.attributes.get(ATTR_DEVICE_CLASS)
+        if not device_class or device_class not in DEVICE_CLASSES:
+            continue
+        entity_ids.append((entity_id, device_class))
+    return entity_ids
+
+
+# Faster than try/except
+# From https://stackoverflow.com/a/23639915
+def _is_number(s: str) -> bool:
+    """Return True if string is a number."""
+    return s.replace(".", "", 1).isdigit()
+
+
+def compile_statistics(
+    hass: HomeAssistant, start: datetime.datetime, end: datetime.datetime
+) -> dict:
+    """Compile statistics for all entities during start-end.
+
+    Note: This will query the database and must not be run in the event loop
+    """
+    result: dict = {}
+
+    entities = run_callback_threadsafe(hass.loop, _get_entities, hass).result()
```

Comment:
Suggestion above to let `.all` do the callback and call `_get_entities` directly

Reply:
We could also write a threadsafe version of `.all` that avoids the need for a callback

Reply:
```
diff --git a/homeassistant/core.py b/homeassistant/core.py
index c22526474a..bb28ae0479 100644
--- a/homeassistant/core.py
+++ b/homeassistant/core.py
@@ -1049,9 +1049,15 @@ class StateMachine:
 
     def all(self, domain_filter: str | Iterable | None = None) -> list[State]:
         """Create a list of all states."""
-        return run_callback_threadsafe(
-            self._loop, self.async_all, domain_filter
-        ).result()
+        if domain_filter is None:
+            return list(self._states.values())
+
+        if isinstance(domain_filter, str):
+            domain_filter = (domain_filter.lower(),)
+
+        return [
+            state for state in list(self._states.values()) if state.domain in domain_filter
+        ]
 
     @callback
     def async_all(self, domain_filter: str | Iterable | None = None) -> list[State]
```

Reply:
How is this threadsafe?

Reply:
Fixed in 11542984aa06bd43dfe49112bbf1d92e6b51297e